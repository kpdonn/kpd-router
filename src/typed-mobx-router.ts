import { History } from 'history'
import { Diff } from 'type-utils'
import { pb } from 'router-impl'

export interface RouterBuilder<T> {
  start(): T & { routerStore: RouterStore }

  addRoute<
    N extends string,
    R extends string,
    L extends ((params: OnLoadParams<R, Q, CN, CT, D>) => void),
    Q extends string,
    CN extends string,
    CT,
    D extends Defaults<R, Q, CN, CT>,
    Extra
  >(route: {
    name: N
    path: [string, R[]]
    queryParams?: Q[]
    onLoad?: L & Extra
    defaults?: D
    converters?: Converters<CN, CT>
  }): RouterBuilder<
    T & Record<N, (args: GoToRouteParams<R, Q, CN, CT, D>) => void>
  >
}

export type GoToRouteParams<
  R extends string,
  Q extends string,
  CN extends string,
  CT,
  D extends Defaults<R, Q, CN, CT>
> = Params<Diff<Diff<R, keyof D>, CN>, Diff<Q | keyof D, CN>> &
  Record<Diff<CN, Q> & Diff<R, keyof D>, CT> &
  Partial<Record<Diff<CN, Diff<R, keyof D>>, CT>>

export type Defaults<
  R extends string,
  Q extends string,
  CN extends string,
  CT
> = Partial<
  Record<Diff<R, CN>, string> & Record<Diff<Q, CN>, string> & Record<CN, CT>
>

export type OnLoadParams<
  R extends string,
  Q extends string,
  CN extends string,
  CT,
  D extends Defaults<R, Q, CN, CT>
> = Params<Diff<R | keyof D, CN>, Diff<Diff<Q, keyof D>, CN>> &
  Record<Diff<CN, Diff<Q, keyof D>> & R | keyof D, CT> &
  Partial<Record<Diff<CN, R | keyof D> & Diff<Q, keyof D>, CT>>

export interface Converters<CNA extends string, CTA> {
  [0]?: Converter<CNA, CTA>

  [index: number]: Converter<any, any> | undefined
}

export interface Converter<N extends string, T> {
  names: N
  from: (arg: T) => string
}

export type Params<R extends string, Q extends string> = { [K in R]: string } &
  { [K in Q]?: string }

export interface RouterStore {
  routeName: string

  params: object

  readonly currentPath: string
}

const nr = newRouter({} as any)
  .addRoute({
    name: 'test',
    path: pb`/test/${'id'} ${'other'}`,
    queryParams: ['hello'],
    onLoad: args => args,
    defaults: { other: '' },
    converters: [{ names: 'id', from: (nid: number) => nid.toString() }]
  })
  .start()

nr.test({ id: 3 })

declare function reqstring(arg: string): void

export interface Route<N, R, L, Q, D, CN, CT> {
  name: N
  path: [string, R[]]
  queryParams?: Q[]
  onLoad?: L
  defaults?: D
  converter?: [CN[], (arg: CT) => string]
}

export function newRouter(history: History): RouterBuilder<{}> {
  // return new RouterBuilderImpl(history)
  return {} as any
}

export interface PathTemplate {
  <T extends string>(literals: TemplateStringsArray, ...args: T[]): [
    string,
    T[]
  ]
}
