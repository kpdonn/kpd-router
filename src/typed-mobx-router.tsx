import { History } from "history"
import { ReactComponentCreator, RouterBuilder, RouterStore } from "interfaces"
import * as React from "react"
import { autorun, computed, observable, runInAction } from "mobx"
import { RouteManager } from "route-manager"
import { observer } from "mobx-react"

export const path = <T extends string>(
  literals: TemplateStringsArray,
  ...args: T[]
): [string, T[]] => {
  const builtPath = literals
    .map((litString, i) => (args[i] ? litString + ":" + args[i] : litString))
    .join("")

  return [builtPath, args]
}

export function newRouter(history: History): RouterBuilder {
  return new RouterBuilderImpl(history) as any
}

class RouterStoreImpl
  implements RouterStore<{ [index: string]: Function }, any, { route: string; params: any }> {
  goTo: { [index: string]: Function } = {}
  Link: React.ComponentClass<any>

  @observable currentRoute: { route: string; params: any }

  constructor(
    private routeManager: RouteManager,
    readonly history: History,
    readonly componentMap: Map<string, ReactComponentCreator<any>>
  ) {}

  @computed
  get currentPath(): string {
    return this.routeManager.buildRoute(this.currentRoute.route, this.currentRoute.params)
  }
}

class RouterBuilderImpl {
  private routerStore: RouterStoreImpl

  private onLoadMap = new Map<string, (params: any) => void>()

  private routeManager = new RouteManager()
  private componentMap = new Map<string, ReactComponentCreator<any>>()

  constructor(readonly history: History) {
    this.routerStore = new RouterStoreImpl(this.routeManager, this.history, this.componentMap)
  }

  setRouteInfoFromLocation() {
    const match = this.routeManager.matchRoute(
      this.history.location.pathname,
      this.history.location.search
    )

    this.routerStore.currentRoute = { route: match.name, params: match.params }
  }

  start(): RouterStoreImpl {
    this.setRouteInfoFromLocation()

    autorun(() => {
      const { route } = this.routerStore.currentRoute
      const onLoadFun = this.onLoadMap.get(route)
      if (onLoadFun) {
        const params = this.routerStore.currentRoute.params
        onLoadFun(params)
      }
    })

    autorun(() => {
      const path = this.routerStore.currentPath
      if (path !== this.history.location.pathname + this.history.location.search) {
        this.history.push(path)
      }
    })

    return this.routerStore
  }

  addRoute(route: Route): this {
    this.routeManager.addRoute(route)

    const defaults = route.defaults || {}

    this.routerStore.goTo[route.name] = (args: any) => {
      runInAction(() => {
        this.routerStore.currentRoute = { route: route.name, params: { ...defaults, ...args } }
      })
    }

    if (route.component) {
      this.componentMap.set(route.name, route.component)
    }

    if (route.onLoad) {
      this.onLoadMap.set(route.name, route.onLoad)
    }
    return this
  }
}

export interface Route {
  name: string
  path: [string, string[]] | string
  queryParams?: string[]
  onLoad?: (params: any) => void
  defaults?: { [index: string]: any }
  converters?: {
    names: string[]
    toString: (arg: any) => string
    fromString: (arg: string) => any
  }[]
  component?: React.ComponentType
}

@observer
export class Router extends React.Component<{ router: RouterStore<any, any, any> }> {
  render() {
    const routerStoreImpl = this.props.router as RouterStoreImpl

    const { route, params } = this.props.router.currentRoute

    const Component = routerStoreImpl.componentMap.get(route)
    console.log("render component: ", Component)

    if (Component) {
      return <Component {...params} />
    } else {
      return null
    }
  }
}
