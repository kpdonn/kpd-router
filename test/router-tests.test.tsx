import * as raf from "raf"
raf.polyfill(global)

import createMemoryHistory from "history/createMemoryHistory"
import { History } from "history"

import { newRouter, path, Router } from "typed-mobx-router"
import * as React from "react"
import * as TestRenderer from "react-test-renderer"
import { createRenderer } from "react-test-renderer/shallow"
const Main = (a: any) => <span>Main</span>
const PersonList = (a: any) => <div>PersonList</div>
const Person = (a: any) => <div>Person</div>
const numConverter = {
  toString: (id: number) => id.toString(),
  fromString: (arg: string) => Number.parseInt(arg)
}

describe("is on main page to start", () => {
  const mainOnLoad = jest.fn()
  const personListOnLoad = jest.fn()
  const personOnLoad = jest.fn()

  const router = createRouter("/", { mainOnLoad, personListOnLoad, personOnLoad })
  const renderer = createRenderer()
  renderer.render(<Router router={router} />)

  it("current route main", () => expect(router.currentRoute.route).toBe("main"))
  it("mainOnLoad called", () => expect(mainOnLoad).toBeCalled())

  it("renderer should match snapshot", () => expect(renderer.getRenderOutput()).toMatchSnapshot())
})

describe("goes to people when function called", () => {
  const mainOnLoad = jest.fn()
  const personListOnLoad = jest.fn()
  const personOnLoad = jest.fn()

  const router = createRouter("/", { mainOnLoad, personListOnLoad, personOnLoad })
  const renderer = createRenderer()
  renderer.render(<Router router={router} />)
  const history: History = (router as any).history

  router.goTo.personList({})

  it("current route personList", () => expect(router.currentRoute.route).toBe("personList"))
  it("correct params", () => expect(router.currentRoute.params).toEqual({ page: 1 }))

  it("called on load correctly", () => expect(personListOnLoad).toBeCalledWith({ page: 1 }))
  it("correct url", () =>
    expect(history.location.pathname + history.location.search).toBe("/people?page=1"))

  it("renderer should match snapshot", () => expect(renderer.getRenderOutput()).toMatchSnapshot())
})

describe("is on people page 2 to start", () => {
  const mainOnLoad = jest.fn()
  const personListOnLoad = jest.fn()
  const personOnLoad = jest.fn()

  const router = createRouter("/people?page=2", { mainOnLoad, personListOnLoad, personOnLoad })
  const renderer = createRenderer()
  renderer.render(<Router router={router} />)
  const history: History = (router as any).history

  it("current route personList", () => expect(router.currentRoute.route).toBe("personList"))
  it("correct params", () => expect(router.currentRoute.params).toEqual({ page: 2 }))

  it("called on load correctly", () => expect(personListOnLoad).toBeCalledWith({ page: 2 }))
  it("correct url", () =>
    expect(history.location.pathname + history.location.search).toBe("/people?page=2"))

  it("renderer should match snapshot", () => expect(renderer.getRenderOutput()).toMatchSnapshot())
})

describe("goes to person when function called", () => {
  const mainOnLoad = jest.fn()
  const personListOnLoad = jest.fn()
  const personOnLoad = jest.fn()

  const router = createRouter("/", { mainOnLoad, personListOnLoad, personOnLoad })
  const renderer = createRenderer()
  renderer.render(<Router router={router} />)
  const history: History = (router as any).history

  router.goTo.person({ id: "42" })

  it("current route person", () => expect(router.currentRoute.route).toEqual("person"))
  it("correct params", () => expect(router.currentRoute.params).toEqual({ id: "42" }))

  it("called on load correctly", () => expect(personOnLoad).toBeCalledWith({ id: "42" }))
  it("correct url", () =>
    expect(history.location.pathname + history.location.search).toBe("/people/42"))

  it("renderer should match snapshot", () => expect(renderer.getRenderOutput()).toMatchSnapshot())
})

describe("is on person list to start", () => {
  const mainOnLoad = jest.fn()
  const personListOnLoad = jest.fn()
  const personOnLoad = jest.fn()

  const router = createRouter("/people/100", { mainOnLoad, personListOnLoad, personOnLoad })
  const renderer = createRenderer()
  renderer.render(<Router router={router} />)
  const history: History = (router as any).history

  it("current route person", () => expect(router.currentRoute.route).toEqual("person"))
  it("correct params", () => expect(router.currentRoute.params).toEqual({ id: "100" }))

  it("called on load correctly", () => expect(personOnLoad).toBeCalledWith({ id: "100" }))
  it("correct url", () =>
    expect(history.location.pathname + history.location.search).toBe("/people/100"))

  it("renderer should match snapshot", () => expect(renderer.getRenderOutput()).toMatchSnapshot())
})

function createRouter(
  initialPath: string = "/",
  mocks: { mainOnLoad: any; personListOnLoad: any; personOnLoad: any }
) {
  const history = createMemoryHistory({ initialEntries: [initialPath] })
  return newRouter(history)
    .addRoute({
      name: "main",
      path: "/",
      onLoad: mocks.mainOnLoad,
      component: Main
    })
    .addRoute({
      name: "personList",
      path: path`/people`,
      queryParams: ["page"],
      defaults: { page: 1 },
      converters: [{ names: ["page"], ...numConverter }],
      onLoad: mocks.personListOnLoad,
      component: PersonList
    })
    .addRoute({
      name: "person",
      path: path`/people/${"id"}`,
      onLoad: mocks.personOnLoad,
      component: Person
    })
    .start()
}
