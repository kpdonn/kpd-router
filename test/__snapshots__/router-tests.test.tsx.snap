// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`goes to people when function called renderer should match snapshot 1`] = `
<PersonList
  page={1}
/>
`;

exports[`goes to person when function called renderer should match snapshot 1`] = `
<Person
  id="42"
/>
`;

exports[`is on main page to start renderer should match snapshot 1`] = `<Main />`;

exports[`is on people page 2 to start renderer should match snapshot 1`] = `
<PersonList
  page={2}
/>
`;

exports[`is on person list to start renderer should match snapshot 1`] = `
<Person
  id="100"
/>
`;
